lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1248 @@
+            sequences = input
+            step_fun = step
+
+        ones = T.ones((num_batch, 1))
+        if isinstance(self.cell_init, T.TensorVariable):
+            cell_init = self.cell_init
+        else:
+            # Dot against a 1s vector to repeat to shape (num_batch, num_units)
+            cell_init = T.dot(ones, self.cell_init)
+
+        if isinstance(self.hid_init, T.TensorVariable):
+            hid_init = self.hid_init
+        else:
+            # Dot against a 1s vector to repeat to shape (num_batch, num_units)
+            hid_init = T.dot(ones, self.hid_init)
+
+        # The hidden-to-hidden weight matrix is always used in step
+        non_seqs = [self.W_hid_stacked]
+        # The "peephole" weight matrices are only used when self.peepholes=True
+        if self.peepholes:
+            non_seqs += [self.W_cell_to_ingate,
+                         self.W_cell_to_forgetgate,
+                         self.W_cell_to_outgate]
+        # theano.scan only allows for positional arguments, so when
+        # self.peepholes is False, we need to supply fake placeholder arguments
+        # for the three peephole matrices.
+        else:
+            non_seqs += [(), (), ()]
+        # When we aren't precomputing the input outside of scan, we need to
+        # provide the input weights and biases to the step function
+        if not self.precompute_input:
+            non_seqs += [self.W_in_stacked, self.b_stacked]
+        # As above, when we aren't providing these parameters, we need to
+        # supply placehold arguments
+        else:
+            non_seqs += [(), ()]
+
+        if self.unroll_scan:
+            # Explicitly unroll the recurrence instead of using scan
+            cell_out, hid_out = unroll_scan(
+                fn=step_fun,
+                sequences=sequences,
+                outputs_info=[cell_init, hid_init],
+                go_backwards=self.backwards,
+                non_sequences=non_seqs,
+                n_steps=self.input_shape[1])
+        else:
+            # Scan op iterates over first dimension of input and repeatedly
+            # applies the step function
+            cell_out, hid_out = theano.scan(
+                fn=step_fun,
+                sequences=sequences,
+                outputs_info=[cell_init, hid_init],
+                go_backwards=self.backwards,
+                truncate_gradient=self.gradient_steps,
+                non_sequences=non_seqs,
+                strict=True)[0]
+
+        # dimshuffle back to (n_batch, n_time_steps, n_features))
+        hid_out = hid_out.dimshuffle(1, 0, 2)
+        cell_out = cell_out.dimshuffle(1, 0, 2)
+
+        # if scan is backward reverse the output
+        if self.backwards:
+            hid_out = hid_out[:, ::-1]
+            cell_out = cell_out[:, ::-1]
+
+        self.hid_out = hid_out
+        self.cell_out = cell_out
+
+        return hid_out
+
+
+class GRULayer(Layer):
+    r"""Gated Recurrent Unit (GRU) Layer
```

Comment:
It tells Python to ignore escape codes, e.g. the \s in \sigma.  http://stackoverflow.com/a/4780104